README.md
# Vimmy - Neato Vim-ness for Safari

## Commands
- `hjkl` to scroll left/down/up/right
- `gg` and `shift+g` to scroll to top/bottom
- `f` to click on elements like links, inputs, buttons
- `shift-f` to open links in a new background tab
- `esc` and `ctrl+[` to break out of link mode
- `shift-h` to go back in history
- `shift-l` to go forward in history
- `gT` to go to the next tab
- `gt` to go the the previous tab
- `x` to close the current tab
- `ctrl-d` to scroll down by half a page
- `ctrl-u` to scroll up by half a page
- `ctrl-f` to scroll down by a full page
- `ctrl-b` to scroll up by a full page


## Settings

If you use websites that have their own shortcuts, you can add them to the extension blacklist (available since version `0.4.0`). In extension settings, just add a list of domains where you want the extension to be ignored. Each line can be a regular expression. Please separate the values using commas. For example, yours might look something like this:

```
google.com,facebook.com,tumblr.com/.*
```

Unfortunately Safari is quite limited in what kinds of settings can be added at the moment, so I couldn't add a proper list input. Maybe one day!

It's also possible to add a global hard-coded website blacklist. If you have thoughts or suggestions for this, please check out the [Github repo](https://github.com/gggritso/Vimmy.safariextension/issues).

## Why

There are other Vim extensions for Safari, but this one is mine!

- better hints (high contrast, chevrons show associated link)
- fluid scrolling animations
- minimum useful command set


CHANGELOG.md
## 0.5.0

### New Features
- `shift-h` and `shift-l` go back and forward in history respectively

### Improvements
- link that wrap onto multiple lines are now hinted correctly

## 0.4.0

### New Features

- URL blacklisting

### Improvements
- better handling of links at the top of pages
- fix for a scrolling bug in Safari 10
- hints now scroll when the site scrolls

## 0.3.0

### New Features

- `gt` and `gT` commands that go to the next tab, and previous tab respectively

## 0.2.0

### Improvements

- hints are now tilted 20˚ to prevent inconvenient overlap
- activated hints are highlighted with green
- various animations and small design tweaks

### New Features

- shift+f turns on "new tab mode"
- buttons, inputs, and select element can be activated with the keyboard

## 0.1.0

Initial release.


vimmy.js
(function() {
  'use strict';

  var Vimmy = {
    'characters': 'asdfejklio'.toUpperCase().split(''), // characters used for link hints
    'scrollDistance': 150,  // pixel distance for hjkl scrolling
    'scrollDelay': 100, // time to animate scrolling
    'mode': 'command',  // start in command mode, other mode is 'hint'
    'scrollLock': false, // keep track of scrolling, don't allow mashing/holding

    'isBlacklisted': false,
    'hostBlacklist': [ 'facebook.com', 'twitter.com', 'www.tumblr.com', 'getprismatic.com',
      'app.asana.com' ],

    'hintTemplate': function( hint ) {

      return '<span><b>' +  hint.split('').join('</b><b>') + '</b></span>';
    },

    'hintLookup': {}, // all the hints that we've stored, by hint
    'typedCharacters': [], // a place to save the characters a user's typed out for link hints

    'previousKeyName': undefined,  // keep track of the previous keyPress for combos like gg
    'keyLookup': {
      9: 'tab', 17: 'ctrl', 16: 'shift',
      18: 'alt', 27: 'esc', 8: 'backspace',

      65: 'a', 66: 'b', 67: 'c', 68: 'd', 69: 'e',
      70: 'f', 71: 'g', 72: 'h', 73: 'i', 74: 'j',
      75: 'k', 76: 'l', 77: 'm', 78: 'n', 79: 'o',
      80: 'p', 81: 'q', 82: 'r', 83: 's', 84: 't',
      85: 'u', 86: 'v', 87: 'w', 88: 'x', 89: 'y',
      90: 'z',

      219: '[', 191: '/'

    },

    'scroll': function( x, y, relative ) {
      // Scroll to a position with a quick jQuery animation
      if ( Vimmy.scrollLock ) return;

      Vimmy.scrollLock = true;

      var modifier = ( relative ) ? '+=' : '',
        animateScroll = function( props ) {
          $( 'body' ).animate( props, Vimmy.scrollDelay, function() {
            Vimmy.scrollLock = false; // clear the lock on animation end
          });
        };

      // animate the transition
      if ( y !== null ) {
        animateScroll({ 'scrollTop': modifier + y + 'px' });
      }

      if ( x !== null ) {
        animateScroll({ 'scrollLeft': modifier + x + 'px' });
      }

    },

    'followLink': function( $link ) {

      var href = $link.attr( 'href' ),
        destination;

      if ( $link.attr( 'target' ) === '_blank' ) {
        destination =  'new';
      } else {
        destination = 'same' ;
      }

      // Clean up the href
      if ( href.substring( 0, 4 ) !== 'http' ) { // i.e. it's a relative link
        if ( href.substring( 0, 1) === '/' ) { // i.e. it's to the root
          href = window.location.origin + href;
        } else { // i.e. it's from the same folder
          href = window.location.href.match(/.+\//) + href;
        }
      }

      // Follow the white rabbit with a slight UI delay
      setTimeout( function() {
        if ( destination === 'same' ) {
          window.location.href = href;
        } else if ( destination === 'new' ) {
          safari.self.tab.dispatchMessage( 'newtab', href );
        }

        Vimmy.typedCharacters = [];
        $( '#vimmyHints span' ).show();
      }, 100 );

      return true;
    },

    'processKey': function( event ) {
      // Takes a keypress event and returns a nice, clean keyName

      var keyCode = event.keyCode,
        keyName = '',
        modifier = '';

      // Listen for modifiers, makes things a bit easier
      // There are a lot of extras here, but I'm leaving them in case I need some more
      if ( event.metaKey ) {
        modifier += 'cmd-';
      }
      if ( event.ctrlKey ) {
        modifier += 'ctrl-';
      }
      if ( event.altKey ) {
        modifier += 'alt-';
      }
      if ( event.shiftKey ) {
        modifier += 'shift-';
      }

      // If the key is a number, assign its name quickly, otherwise do a lookup
      if ( keyCode >= 48 && keyCode <= 57 ) {
        keyName = ( keyCode - 48 ).toString();
      } else {
        keyName = Vimmy.keyLookup[ keyCode ];
      }

      // Don't double up on the modifiers, store only if it's a meaningful keypress
      if ( $.inArray( keyCode, [16, 17, 18, 91] ) < 0 ) {
        keyName = modifier + keyName;
      }

      return keyName;
    },

    'showHints': function( ) {
      // Display all the nice little key hints!
      var links = $( 'a:in-viewport:visible' ),
        linkCount = links.length,
        charCount = Vimmy.characters.length,
        hintLength = Math.ceil( Math.log( linkCount ) / Math.log( charCount ) ),
        hints = [],
        i, j, c, hint, $hint, $link, linkOffset;

      if ( linkCount > 0 && hintLength === 0 ) {
        // Math, bro - log(1) === 0 :(
        hintLength = 1;
      }

      // Create the hint text snippets AAA, AAB, ACD, etc
      $( 'body' ).append( $( '<div id="vimmyHints"></div>' ) );
      for ( i = 0; i < linkCount; i++ ) {
        // for every link
        hint = '';
        $link = $( links[ i ] );

        for ( j = 1; j <= hintLength; j++ ) {
          // for every character that we need to get until we have a long enough hint
          c = Math.floor( i /  Math.pow( charCount, hintLength - j ) ) % charCount;
          hint += Vimmy.characters[ c ];
        }

        $hint = $( Vimmy.hintTemplate( hint ) );

        // position the hint nicely
        linkOffset = $link.offset();
        linkOffset.left -= ( 23 + hintLength * 5 );
        linkOffset.top += ( -9 + ( $link.height() / 2 ) );
        $hint.offset( linkOffset );

        hints.push( $hint );
        Vimmy.hintLookup[ hint ] = { '$hint': $hint, '$link': $link };

      }

      $( '#vimmyHints' ).append( hints ); // append everything in one shot

    },

    'hideHints': function() {
      $( '#vimmyHints' ).remove(); // since they're generated for only chunks of pages, it's easier
                                   // to just remove them completely
    },

    'checkHint': function() {
      // Process the hint. If it's incomplete, highlight what's the good news, if it's complete
      // trigger the link

      $( '#vimmyHints b' ).removeClass( 'typed' );
      $( '#vimmyHints span' ).show();

      var sequence = Vimmy.typedCharacters.join('').toUpperCase(),
        hint, $hint, $link;

      for ( hint in Vimmy.hintLookup ) {
        if ( Vimmy.hintLookup.hasOwnProperty( hint ) ) {
          $link = Vimmy.hintLookup[ hint ].$link; // i.e. the hint properties, $link and $hint
          $hint = Vimmy.hintLookup[ hint ].$hint; // i.e. the hint properties, $link and $hint

          // Highlight the hint, hide all irrelevant ones
          if ( hint.substring(0, sequence.length) === sequence) {
            $hint.children().slice(0, sequence.length ).addClass( 'typed' );
          } else {
            $hint.hide();
          }

          if ( hint === sequence ) {
            Vimmy.followLink( $link );

            Vimmy.typedCharacters = [];
            $( '#vimmyHints span' ).show();
            break;

          }
        }
      }

    },

    'keyHandler': function( event ) {

      // If user is inside an input, ignore all actions
      if ( $.inArray( document.activeElement.tagName.toUpperCase(), ['INPUT', 'TEXTAREA'] ) > -1 ) {
        return;
      }

      // If it's a box with contenteditable (basically an input), ignore actions
      if ( $( document.activeElement ).attr( 'contenteditable' ) === 'true' ) {
        return;
      }


      var keyCode = event.keyCode,
        keyName = Vimmy.processKey( event );

      if ( Vimmy.mode === 'command' ) {
        // The simple case, just listen for keybindings

        if ( keyName === 'f' ) {
          Vimmy.mode = 'hint'; // enter hint mode for new tab opening
          Vimmy.showHints();
        }

        // Basic hjkl scrolling (not enabled on all sites!)
        if ( !Vimmy.isBlacklisted ) {
          if ( keyName === 'j' ) {
            Vimmy.scroll( null, Vimmy.scrollDistance, true );
          } else if ( keyName === 'k' ) {
            Vimmy.scroll( null, -Vimmy.scrollDistance, true );
          } else if ( keyName === 'h' ) {
            Vimmy.scroll( -Vimmy.scrollDistance, null, true );
          } else if ( keyName === 'l' ) {
            Vimmy.scroll( Vimmy.scrollDistance, null, true );
          }
        }

        // gg, G
        if ( keyName === 'g' && Vimmy.previousKeyName === 'g' ) {
          Vimmy.scroll( null, 0 );
        }
        if ( keyName === 'shift-g' ) {
          Vimmy.scroll( null, $( document ).height() );
        }

      } else { // Vimmy.mode === 'hint'
        // We're typing in links right now and such

        // Cut out of hint mode with escape of ctrl+[
        if ( keyName === 'esc' || keyName === 'ctrl-[' ) {
          Vimmy.mode = 'command';
          Vimmy.typedCharacters = [];
          Vimmy.hideHints();
          return;  // We don't need to do anything else here
        }

        // Backspace - remove last typed character
        if ( keyName === 'backspace' ) {
          Vimmy.typedCharacters.pop();
          Vimmy.checkHint();
          return;
        }

        // All's okay, nothing crazy, process the input to see if we're ready
        if ( keyName && $.inArray( keyName.toUpperCase(), Vimmy.characters ) > -1 ) {
          // Only listen to characters that compose the hints
          Vimmy.typedCharacters.push( keyName );
          Vimmy.checkHint();
        }
      }

      // Don't save the previousKey if it's a modifier
      if ( Vimmy.mode === 'command' && $.inArray( keyCode, [16, 17, 18, 91] ) === -1 ) {
        Vimmy.previousKeyName = keyName;
      }

    }
  };

  // Add the global key press listener
  // Do not bind (or do anything) to iframes! This caused a ton of issued related to Facebook "like"
  // buttons and other such embedded objects
  if ( top === self ) {
    $( window ).bind( 'keydown', Vimmy.keyHandler );

    // Add browser events listener (come from the extension HTML page)
    // Check the blacklist
    $.each( Vimmy.hostBlacklist, function( i, blockedHost ) {
      if ( window.location.host.substring( 0, blockedHost.length ) === blockedHost ) {
        Vimmy.isBlacklisted = true;
      }
    });

  }

})();


src/vimmy2.js
( function() {
  'use strict';

  var
    $window = $( window ),
    $body = $( 'body' );

  var
    // TODO: These should be settings
    CHARACTERS = 'asdfewjklio'.toUpperCase().split( '' ),
    BUMP_DISTANCE = 100, // px
    SCROLL_DURATION = 100;  // ms

  var GREEDY_INPUT_TYPES = [ 'text', 'password', 'phone', 'email', 'search' ];

  var
    TARGETABLE_ELEMENTS = [ 'a', 'button', 'input', 'select' ],
    TARGETABLE_ELEMENT_SELECTOR = TARGETABLE_ELEMENTS.join( ', ' );

  var
    URL_IS_BLACKLISTED = false;

  var KEYCODE_LOOKUP = {
    9: 'tab', 17: 'ctrl', 16: 'shift',
    18: 'alt', 27: 'esc', 8: 'delete',
    91: 'cmd', 93: 'cmd', 13: 'return',

    65: 'a', 66: 'b', 67: 'c', 68: 'd', 69: 'e',
    70: 'f', 71: 'g', 72: 'h', 73: 'i', 74: 'j',
    75: 'k', 76: 'l', 77: 'm', 78: 'n', 79: 'o',
    80: 'p', 81: 'q', 82: 'r', 83: 's', 84: 't',
    85: 'u', 86: 'v', 87: 'w', 88: 'x', 89: 'y',
    90: 'z',

    219: '[', 191: '/'
  };

  // TODO: Benchmark and improve hint creation time

  var MODIFIER_KEYCODES = [ 16, 17, 18, 91, 93 ];

  var
    MODE = 'command',
    FORCE_NEW_TAB = false,

    SCROLL_IS_LOCKED = false,

    HINTS = [],

    PREVIOUS_KEY = null,
    TYPED_HINT_CHARACTERS = [];


  // // // // //

  function boot() {

    if ( amAnIframe() ) return;

    window.addEventListener( 'keydown', vimmyKeyDownHandler, true )
    $body.append( '<div id="vimmy-hints"></div>' );

    safari.self.addEventListener( 'message', function handleMessage( event ) {
      if ( event.name === 'newBlackListedURLs' ) checkBlackListStatus( event.message );
    });

    safari.self.tab.dispatchMessage( 'ready' );
  }


  function amAnIframe() {
    return top !== self;
  }


  function checkBlackListStatus( blackListedURLs ) {

    URL_IS_BLACKLISTED = false;

    var
      i, expression;

    for ( i = 0; i < blackListedURLs.length; i += 1 ) {
      expression = new RegExp( blackListedURLs[ i ].trim() );
      if ( expression.test( window.location.href ) ) URL_IS_BLACKLISTED = true;
    }

  }


  function vimmyKeyDownHandler( event ) {

    if ( elementCapturesKeys( document.activeElement ) ) return;
    if ( URL_IS_BLACKLISTED ) return;

    var
      key = getKeyNameFromEvent( event ),
      $elements;

    if ( MODE === 'command' ) {

      if ( key.isIn([ 'f', 'shift-f' ]) ) {
        swallowEvent( event );

        FORCE_NEW_TAB = ( key === 'shift-f' );

        $elements = getVisibleElements();
        if ( $elements.length === 0 ) return;

        MODE = 'elements';

        HINTS = createHintsForElements( $elements );
        showHints();

        return;
      }

      if ( [ 'h', 'j', 'k', 'l' ].contains( key ) ) {
        swallowEvent( event );

        if ( key === 'h' ) scrollBy( -BUMP_DISTANCE, 0 );
        if ( key === 'j' ) scrollBy( 0, BUMP_DISTANCE );
        if ( key === 'k' ) scrollBy( 0, -BUMP_DISTANCE );
        if ( key === 'l' ) scrollBy( BUMP_DISTANCE, 0 );

        return;
      }

      if ( key === 'ctrl-u' ) {
        swallowEvent( event );

        scrollBy( 0, -( window.innerHeight / 2 ) );
      }

      if ( key === 'ctrl-d' ) {
        swallowEvent( event );

        scrollBy( 0, ( window.innerHeight ) / 2 );
      }

      if ( key === 'ctrl-b' ) {
        swallowEvent( event );

        scrollBy( 0, -window.innerHeight );
      }

      if ( key === 'ctrl-f' ) {
        swallowEvent( event );

        scrollBy( 0, window.innerHeight );
      }

      if ( key === 'g' ) {
        swallowEvent( event )
      }

      if ( key === 'g' && PREVIOUS_KEY === 'g' ) scrollTo( null, 0 );
      if ( key === 'shift-g' ) scrollTo( null, $( document ).height() );

      if ( key === 'shift-t' && PREVIOUS_KEY === 'g' ) goToPreviousTab();
      if ( key === 't' && PREVIOUS_KEY === 'g' ) goToNextTab();

      if ( key === 'shift-h' ) goBackHistory();
      if ( key === 'shift-l' ) goForwardHistory();

      if ( ![ 'cmd', 'esc', 'shift', 'alt', 'tab' ].contains( key ) ) PREVIOUS_KEY = key;

    } else if ( MODE === 'elements' ) {

      if ( [ 'esc', 'ctrl+[' ].contains( key ) ) {
        swallowEvent( event );

        MODE = 'command';
        hideHints();

        return;
      }

      if ( key === 'delete' ) {
        swallowEvent( event );

        TYPED_HINT_CHARACTERS.pop();
        filterHints( TYPED_HINT_CHARACTERS );

        return;
      }

      if ( key.toUpperCase().isIn( CHARACTERS ) ) {
        swallowEvent( event );

        TYPED_HINT_CHARACTERS.push( key );
        filterHints( TYPED_HINT_CHARACTERS );
      }

    }

  }


  function swallowEvent( event ) {

    event.preventDefault();
    event.stopPropagation();
    event.stopImmediatePropagation();
  }


  function elementCapturesKeys( element ) {

    var
      tag = element.tagName,
      type = element.getAttribute( 'type' );

    if ( tag === 'TEXTAREA' ) return true;
    if ( tag === 'INPUT' && type.isIn( GREEDY_INPUT_TYPES ) ) return true;
    if ( element.contentEditable === 'true' ) return true;


    return false;
  }


  function getKeyNameFromEvent( event ) {

    var
      keyName,
      modifier = '';

    if ( event.metaKey ) modifier += 'cmd-';
    if ( event.ctrlKey ) modifier += 'ctrl-';
    if ( event.altKey ) modifier +=  'alt-';
    if ( event.shiftKey ) modifier += 'shift-';

    if ( event.keyCode >= 48 && event.keyCode <= 57 ) {
      keyName = ( event.keyCode - 48 ).toString();
    } else {
      keyName = KEYCODE_LOOKUP[ event.keyCode ];
    }

    if ( MODIFIER_KEYCODES.contains( event.keyCode ) ) return keyName;
    return modifier + keyName;
  }


  function getVisibleElements() {

    return Array.prototype.slice
      .call( document.querySelectorAll( TARGETABLE_ELEMENT_SELECTOR ) )
      .filter( inViewPort )
      .filter( isVisible );

  }


  function inViewPort( element ) {

    var bounds = element.getBoundingClientRect()

    return ( bounds.left >= 0 && bounds.top >= 0 && bounds.bottom <= window.innerHeight && bounds.right <= window.innerWidth );

  }


  function isVisible( element ) {
    // TODO: Check performance issues
    if ( element.offsetWidth === 0 && element.offsetHeight === 0 ) return false;
    if ( element.getClientRects().length === 0 ) return false;

    var style = window.getComputedStyle( element );

    if ( style.visibility === 'hidden' ) return false;
    if ( style.opacity === '0' ) return false;

    return true;
  }


  function createHintsForElements( $elements ) {

    var
      requiredHintLength = Math.floor( Math.log( $elements.length ) / Math.log( CHARACTERS.length ) ) + 1, // TODO: Abstract out
      j = 0, i = 0,
      text;

    return $elements.map( function( $element ) {

      text = '';
      for ( j = 1; j <= requiredHintLength; j++ ) {
        text += CHARACTERS[ Math.floor( i /  Math.pow( CHARACTERS.length, requiredHintLength - j ) ) % CHARACTERS.length ]; // TODO: Abstract out
      }

      i += 1;

      text = text.reverse(); // A usability feature, ensure that adjacent hints start with different letters

      return {
        text: text,
        $hint: makeHintHTML( $element, text ),
        $element: $element,
      };

    });

  }


  function makeHintHTML( $element, text ) {

    // TODO: Attach based on the top of the element, skip the extra calculations!
    var
      bounds = $element.getClientRects()[ 0 ],
      attachment = ( bounds.left > 40 && bounds.top > 10 ) ? 'left' : 'right';

    var
      left,
      top;

    if ( attachment === 'left' ) {
      left = ( bounds.left - ( ( text.length * 8 ) + 14 ) ),
      top = ( bounds.top + bounds.bottom ) / 2 - 15;
    } else {
      left = ( bounds.right ) + 8,
      top = ( bounds.top + bounds.bottom ) / 2 - 1;
    };

    return $( '<span class="' + attachment + '" style="left: ' + left + 'px; top: ' + top + 'px;">' +
      '<b>' +  text.split('').join('</b><b>') +
      '</b></span>' );

  }


  function hideHints() {

    TYPED_HINT_CHARACTERS = [];

    $( '#vimmy-hints' )
      .html( '' );

  }


  function showHints() {

    var $hints = HINTS.map( function( hint ) {
      return hint.$hint;
    });

    $( '#vimmy-hints' )
      .css({ top: window.scrollY })
      .append( $hints );

  }


  function filterHints( characters ) {

    var
      prefix = characters.join( '' ).toUpperCase(),
      activeHints = [],
      activeHint;

    HINTS.forEach( function checkHint( hint ) {

      if ( hint.text.startsWith( prefix ) ) {
        // This hint is valid!
        activeHints.push( hint );
        hint.$hint.children()
          .removeClass( 'typed' )
          .slice( 0, prefix.length )
          .addClass( 'typed' );

        hint.$hint.removeClass( 'ineligible' );
      } else {
        hint.$hint.addClass( 'ineligible')
      }

    });

    if ( activeHints.length > 1 ) return;
    activeHint = activeHints[ 0 ];

    if ( activeHint.text === prefix ) {
      activateElement( activeHint.$element );
      highlightHint( activeHint.$hint );
    }

  }


  function activateElement( $element ) {

    var activator;

    if ( isLinkLike( $element ) ) {
      activator = activateAnchor;
    } else if ( isButtonLike( $element ) ) {
      activator = clickElement;
    } else {
      activator = focusElement;
    }


    window.setTimeout( function() {
      activator( $element );
    }, 0 );


    window.setTimeout( function() {

      hideHints();
      MODE = 'command';
    }, 150 );

  }

  function highlightHint( $hint ) {

    $hint
      .addClass( 'accepted' )
      .find( 'b' )
      .removeClass( 'typed' );

  }


  function isLinkLike( $element ) {
    return $element.tagName === 'A';
  }


  function isButtonLike( $element ) {
    if ( $element.tagName === 'BUTTON' ) return true;
    if ( $element.getAttribute( 'type' ) === 'button' ) return true;

    return false
  }


  function focusElement( $element ) {

    $element.focus();
  }


  function activateAnchor( $anchor ) {

    var
      href = $anchor.getAttribute( 'href' ),
      target = $anchor.getAttribute( 'target' ),
      url = makeAbsoluteUrl( href );

    if ( FORCE_NEW_TAB ) {

      safari.self.tab.dispatchMessage( 'newtab', {
        url: url,
        destination: 'background',
      });

    } else if ( target === '_blank' ) {

      safari.self.tab.dispatchMessage( 'newtab', {
        url: url,
        destination: null,
      });

    } else {

      clickElement( $anchor );
    }

  }


  function clickElement( $element ) {

    $element.click();
  }


  function makeAbsoluteUrl( url ) {

    if ( url.startsWith( '#' ) ) {
      return window.location.protocol + '//' + window.location.host + window.location.pathname + window.location.search + url;
    }

    if ( !url.startsWith( 'http' ) ) {

      if ( url.startsWith( '//' ) ) {
        url = window.location.protocol + url;
      } else if ( url.startsWith( '/' ) ) {
        url = window.location.origin + url;
      } else {
        url = window.location.href.match(/.+\//) + url;
      }

    }

    return url;
  }


  function scrollBy( x, y ) {
    scroll( '+=' + ( x || 0 ) , '+=' + ( y || 0 ) );
  }


  function scrollTo( x, y ) {
    scroll( x, y );
  }


  function scroll( left, top ) {

    if ( SCROLL_IS_LOCKED ) return;
    SCROLL_IS_LOCKED = true;

    var animationProperties = {};

    if ( top !== null ) animationProperties.scrollTop = top + 'px';
    if ( left !== null ) animationProperties.scrollLeft = left + 'px';

    $body.animate( animationProperties, SCROLL_DURATION, function() {
      SCROLL_IS_LOCKED = false;
    });

  }


  function goToNextTab() {
    safari.self.tab.dispatchMessage( 'nextTab' );
  }


  function goToPreviousTab() {
    safari.self.tab.dispatchMessage( 'previousTab' );
  }

  function goBackHistory() {
    window.history.back();
  }

  function goForwardHistory() {
    window.history.forward();
  }

  // // // // //

  boot();

})();
